rov/movement/thruster/Force_Limiter.py

```diff
@@ -0,0 +1,117 @@
+    2. acceptable_individual_total
+    3. acceptable_total_change
+    4. acceptable_individual_change
+    """
+
+    def convertToPowerUnit(self, an8vector):
+        # equation: PWM vs Power = y = 0.00114535x^2-3.4372x+2579.273
+        #for x in an8vector:
+            #do something
+
+        return 1
+
+
+    # TODO: edit powerThresh constant, import thruster_power function function, edit else return value
+    def acceptable_total_power(self):
+        # tests the total power consumed
+
+        POWER_THRESH = 1000
+
+        totalPower = sum(convertToPowerUnit(self.an8vector))
+
+        if totalPower < POWER_THRESH:
+            return 1
+        else:
+            # TODO: JASON: Need function to calculate direction scalar from power scalar.
+            return floor(POWER_THRESH/totalPower * 10) / 10
+
+    # TODO: edit ind_thresh
+    def acceptable_individual_total(self):
+        # tests whether each individual thruster is exceeding the maximum
+        # "Limiting"
+        # should return 1 if everything is acceptable, otherwise a value \in (0,1) which will be multiplied to the 6vector
+
+        IND_THRESH = 300
+
+        maxComp = max(self.an8vector)
+
+        if maxComp < IND_THRESH:
+            return 1
+        else:
+            # TODO: JASON: Need function to calculate direction scalar from power scalar.
+            return floor(IND_THRESH / maxComp * 10) / 10
+
+    # TODO: edit total_change_thresh
+    def acceptable_total_change(self, difference8):
+        # tests whether the total "instantaneous" change in power consumed summed among all thrusters is within a threshold
+        # should return 1 if everything is acceptable, otherwise a value \in (0,1) which will be multiplied to the 6vector
+
+        TOTAL_CHANGE_THRESH = 1000
+
+        sumChange = sum(difference8)
+
+        if sumChange < TOTAL_CHANGE_THRESH:
+            return 1
+        else:
+            # TODO: JASON: Need function to calculate direction scalar from power scalar.
+            return floor(TOTAL_CHANGE_THRESH / sumChange * 10) / 10
+
+    # TODO: edit ind_change_thresh
+    def acceptable_individual_change(self, difference8):
+        # tests whether the "instantaneous" change in the thruster value is acceptable for each individual thruster
+        # "Ramping"
+        # should return 1 if everything is acceptable, otherwise a value \in (0,1) which will be multiplied to the 6vector
+
+        IND_CHANGE_THRESH = 1000
+
+        # get greatest absolute value
+        max_change = abs(max(min(difference8), max(difference8), key=abs))
+
+        if max_change < IND_CHANGE_THRESH:
+            return 1
+        else:
+            # TODO: JASON: Need function to calculate direction scalar from power scalar.
+            return floor(IND_CHANGE_THRESH / max_change * 10) / 10
+
+    def enforce(self, a6vector, last8vector, disabledT):
+
+        self.a6vector = a6vector
+        self.an8vector = mapper.calculate(a6vector, disabledT)
```

Comment:
I feel like the `mapper.calculate` should be done outside the enforce calculation and passed into you. I guess this depends on how the whole thruster control system is written.

Reply:
True mapping is supposed to be done after thrust limiting. So its done within here just so we can get their temporary values.

Reply:
I thought the process is:
1) True Mapping
2) Check thrust limiting
3) Readjust forces if not acceptable, go back to step 1
4) Yay, we're done

Reply:
I mean, what's the difference? It's more linear and easy to understand on the high level the current way.

Reply:
Well wait, so you're running thrust mapping in here, modifying the force values, returning the force values, then doing thrust mapping again outside this function?